### PR TITLE
Make Callbacks Easier to Use

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:
Make it easier to use Callbacks: make it useful, and prevent wrong usage. Address https://github.com/palantir/atlasdb/issues/3353

**Implementation Description (bullets)**:
Added a bunch of docs, and added an interface that should make it simpler to create your callback.

**Concerns (what feedback would you like?)**:
Is this actually simpler now? Do we need a better example -- I could have `ConsistencyCheckRunner` implement `CallbackInitializable` instead of being a `Callback`, but not sure it makes sense to do so just to have another example in the codebase.

**Where should we start reviewing?**:
`CallbackInitializable.java`

**Priority (whenever / two weeks / yesterday)**:
Not super urgent

@pkoenig10 for SA

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/3360)

### CODE REVIEW

File: docs/source/miscellaneous/asynchronous-initialization.rst [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+To simplify initialization of objects depending on a transaction manager that may be asynchronously initialized,
+``TransactionManagers`` has an ``asyncInitializationCallback()`` parameter allowing users to specify a custom callback
+that should be executed when ready.
+
+If ``initializeAsync`` is set to false, the callback will be run synchronously (blocking until it is done) as the last step in creating the transaction manager.
+If ``initializeAsync`` is set to true, the callback will also be run asynchronously:
+
+1. Once all the resources for the transaction manager have become ready, **but before the transaction manager becomes initialized**, the registered callback will be run asynchronously.
+2. If the callback fails and should not be retried, the transaction manager will be closed.
+3. If the callback is successful, the transaction manager then becomes initialized.
```
Comment (pkoenig10) [inline docs/source/miscellaneous/asynchronous-initialization.rst]: Probably worth clarifying that the transaction manager provided to the callback can be used before the created transaction manager is initialized.

File: atlasdb-commons/src/main/java/com/palantir/async/initializer/Callback.java [hunk 2]
```diff
@@ -40,7 +52,7 @@
      * @param initException Throwable thrown by init()
      */
     public void cleanup(R resource, Throwable initException) {
-        Throwables.rewrapAndThrowUncheckedException(initException);
+        throw Throwables.rewrapAndThrowUncheckedException(initException);
```
Comment (jeremyk-91) [inline atlasdb-commons/src/main/java/com/palantir/async/initializer/Callback.java]: 👍

File: atlasdb-commons/src/main/java/com/palantir/async/initializer/LambdaCallback.java [hunk 3]
```diff
@@ -16,21 +16,67 @@
...
     }
 
-    public static <R> Callback<R> of(Consumer<R> consumer) {
-        return new LambdaCallback<>(consumer);
+    /**
+     * Creates a callback that will retry initialization until cleanup also throws. If initialize throws,
+     * onInitFailureCleanup is called before calling initialize again.
+     * @param initialize initialization method.
+     * @param onInitFailureCleanup cleanup to be done in case initialization throws. If this method also throws, no more
+     * retries will be attampted and it will bubble up to the caller.
```
Comment (jeremyk-91) [inline atlasdb-commons/src/main/java/com/palantir/async/initializer/LambdaCallback.java]: nit: `attampted`

File: atlasdb-commons/src/test/java/com/palantir/async/initializer/LambdaCallbackTest.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    private void markCleanupThenFailIfMoreThatTenInits(ModifiableInitsAndCleanups initsAndCleanups, Throwable ignore) {
+        markCleanup(initsAndCleanups, ignore);
+        if (initsAndCleanups.inits() > 10) {
+            throw CLEANUP_FAIL;
+        }
+    }
+
+    @Value.Modifiable
```
Comment (jeremyk-91) [inline atlasdb-commons/src/test/java/com/palantir/async/initializer/LambdaCallbackTest.java]: Interesting! Wasn't aware of this feature

File: atlasdb-commons/src/test/java/com/palantir/async/initializer/LambdaCallbackTest.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    public void singleAttemptCallbackCallsInitializeOnceAndSkipsCleanupOnSuccess() {
+        callback = LambdaCallback.singleAttempt(this::markInit, this::markCleanup);
+
+        callback.runWithRetry(initsAndCleanups);
+        assertThat(initsAndCleanups.inits()).isEqualTo(1);
+        assertThat(initsAndCleanups.cleanups()).isEqualTo(0);
+    }
+
+    @Test
+    public void singleAttemptCallbackCallsInitializeAndCleanuoOnceThenRethrowsOnInitFailure() {
```
Comment (jeremyk-91) [inline atlasdb-commons/src/test/java/com/palantir/async/initializer/LambdaCallbackTest.java]: nit: `Cleanup`

File: docs/source/miscellaneous/asynchronous-initialization.rst [hunk 6]
```diff
@@ -0,0 +1,32 @@
+===========================
+Asynchronous Initialization
+===========================
```
Comment (jeremyk-91) [inline docs/source/miscellaneous/asynchronous-initialization.rst]: Thanks for writing this! I don't think we documented this properly before (certainly not in such detail)

File: atlasdb-commons/src/main/java/com/palantir/async/initializer/CallbackInitializable.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+     * Returns a Callback that runs initialize only once. On initialization failure, executes the specified cleanup and
+     * then wraps and throws the cause.
+     */
+    default Callback<R> singleAttemptCallback() {
+        return LambdaCallback.singleAttempt(this::initialize, this::onInitializationFailureCleanup);
+    }
+
+    /**
+     * Returns a Callback that will retry initialization on failure, unless the cleanup task also throws, in which case
+     * the cause is not caught.
```
Comment (jeremyk-91) [inline atlasdb-commons/src/main/java/com/palantir/async/initializer/CallbackInitializable.java]: Probably implied, but for clarity I'd say `in which case the exception thrown by the cleanup task is propagated`

File: atlasdb-commons/src/test/java/com/palantir/async/initializer/LambdaCallbackTest.java (hunk 4 above)
Comment (gmaretic) [reply atlasdb-commons/src/test/java/com/palantir/async/initializer/LambdaCallbackTest.java]: Yeah really useful, I saw it in one of the recent PR's

File: atlasdb-commons/src/main/java/com/palantir/async/initializer/CallbackInitializable.java (hunk 7 above)
Comment (gmaretic) [reply atlasdb-commons/src/main/java/com/palantir/async/initializer/CallbackInitializable.java]: That's the word I was looking for!

**Pull request merged**